### ISSUE TITLE
21 second delay with both node and ffmpeg (Mojave 10.14.6)

### ISSUE THREAD

**Issue by MSteggles** [muammar/mkchromecast#292]
Hi, 
I've just installed mkchromecast. I can get output working with both node and ffmeg, both of which have 21s delay. Having a look it appears this is a problem others have had, but couldn't piece together if there is a definite fix? I've tried switching around sample rates, output formats, backends but it's always 21s.
Cheers

**Comment by muammar** [comment]
You are right. I confirm this is happening for me too. But it is just with Chromecast. When I use my sonos system this delay is below 4 seconds. I honestly do not know how to fix this.

**Comment by fornellas** [comment]
Same for Linux + Chromecast Audio: 21s delay.

At first I thought this was related to Pulseaudio + ffmpeg generating delays. I did some simple test with:

```
parec --format=s16le | ffmpeg -loglevel panic -ac 2 -f s16le -ar 44100 -i - -ac 2 -f s16le -ar 44100 -
```

And the output delay when audio starts / stops is < 1s.

I also tried this:
```
mkfifo i.aac
parec --format=s16le | ffmpeg -loglevel panic -ac 2 -f s16le -ar 44100 -i - -ac 2 -f s16le -ar 44100 > i.aac
mkchromecast -i i.aac
```
with same 20s+ delay.

The curious bit is: when I ^C on mkchromecast, audio stops playing immediately!

My guesses at this point:

- mkchromecast has some 20s buffering before sending to the Chromecast Audio.
- mkchromecast sends data immediately to Chromecast Audio and the 20s delay comes from it.

Chromecast Audio has a "group delay connection" setting. Mine's set to 0s. Perhaps mkchromecast is setting it (or any other delay setting) to 20s (or not setting it, and it is defaulting to 20s).

**Comment by muammar** [comment]
> @muammar I'm thinking that the solution here is to stream data with MPEG-DASH. Do you have any experience with different formats to say if this is a good route?

Good finding. I have no idea! I did this that decrease logging but is just a hack :( 

https://github.com/muammar/mkchromecast/commit/30f4e85bf1220ea5534748adf65f77b35c1b255e

**Comment by cmcinroy** [comment]
Looks like the related issues have been consolidated here. In that vein, just another point of reference:
Mkchromecast v0.3.9, Fedora 33, kernel 5.10.7, GNOME 3.38.3 here.
I was experiencing a delay of 10-20s casting to Chromecast audio.
Setting preferences to ffmpeg/wav/44100 reduced the delay to the 3s range.

**Comment by ojdo** [comment]
> Setting preferences to ffmpeg/wav/44100 reduced the delay to the 3s range.

This is the way. Also, this saves needless re-encoding of the data. Every stable wifi should be able to handle the ~1 Mbit/s stream, which is much for audio but still small compared to any video stream you might pipe through your local setup without issues.

----

Full reference: got my mkchromecast on Arch via `yay mkchromecast` (thanks AUR packagers!) and then launched `mkchromecast -s --encoder-backend ffmpeg -c wav`, selecting my "everywhere" speaker group. Switched to "mkchromecast" in KDE's audio output device settings. 3s delay on all speakers, without any fiddling. Nice!
